core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java

```diff
@@ -251,13 +253,25 @@ public static RelDataType toSql(final RelDataTypeFactory typeFactory,
               type.getFieldNames()),
           type.isNullable());
     } else if (type instanceof JavaType) {
-      return typeFactory.createTypeWithNullability(
-          typeFactory.createSqlType(type.getSqlTypeName()),
-          type.isNullable());
+      Class javaClass = ((JavaType) type).getJavaClass();
+      final RelDataType nonNullType;
+      if (javaClass.isArray()) {
+        nonNullType = typeFactory.createArrayType(type.getComponentType(), -1);
+      } else if (isCollection(javaClass)) {
+        nonNullType = typeFactory.createArrayType(typeFactory.createSqlType(SqlTypeName.ANY), -1);
+      } else {
+        nonNullType = typeFactory.createSqlType(type.getSqlTypeName());
+      }
```

Comment:
Can you reference `JavaToSqlTypeConversionRules` for the type mapping ?

Reply:
There is no need to use ```JavaToSqlTypeConversionRules``` here because ```type``` contains SqlTypeName. Most probably this check just need to be rewritten to switch case based on the enum values and further java class investigation isn't necessary here. To be honest the API of ```typeFactory``` which promises to ```createSqlType(...)``` for sqlTypeName and implicitly throws assertion errors is really bad.

Reply:
I didn't mean to use `JavaToSqlTypeConversionRules` directly, i meant to match all the java collection classes that synced with what `JavaToSqlTypeConversionRules` has already matched.